### PR TITLE
clear up temporary directory on startup and quit

### LINKED ISSUE #2770: Clean up the temporary directory
We use tmp to store exported msgs aswell as downloaded attachments. Windows doesn't automatically clear it out, so we should be clearing it manually.

### PR DESCRIPTION
**Pull request by johnbotris**
We do it at startup because on windows we won't receive before-quit, will-quit or quit events when the user shuts down or logs out without quitting the app first.

this commit also has some consolidation of tmp dir usage and minor tidying up

close #2770

### CODE REVIEW

File: src/desktop/IPC.js [hunk 1]
```diff
@@ -262,27 +261,28 @@ export class IPC {
...
-					              this._wm.get(windowId)?._browserWindow.webContents.startDrag({
-						              files,
-						              icon: this._dragIcon
-					              })
-				              })
+				const getExportPath = async id => path.join(await getExportDirectoryPath(this._electron.app), mailIdToFileName(id, "msg"))
+				const files = await Promise.all(ids.map(getExportPath))
+				                           .then(files => files.filter(fileExists))
+
+				this._wm.get(windowId)?._browserWindow.webContents.startDrag({
```
Comment (charlag) [inline src/desktop/IPC.js:280]: ooohohoho question mark operator? noope we are not this cool yet

**Comment by charlag** [comment]
We have "clarFileData" command for mobile, maybe it's best to be consistent?

File: src/desktop/IPC.js (hunk 1 above)
Comment (johnbotris) [reply src/desktop/IPC.js:280]: :cry: ok I'll update it

**Comment by johnbotris** [comment]
> We have "clarFileData" command for mobile, maybe it's best to be consistent?

Oh yeah, we can just run the cleanup when that gets called normally. Should we still do cleanup on app close? Maybe best to just do it at startup like we already do on mobile

**Comment by charlag** [comment]
I think we might do more cleanups because on desktop folder isolation is not as good as on mobile but we should at least do it when web part asks for it. Just need to make sure that web part does when env is desktop

File: src/login/LoginViewController.js [hunk 2]
```diff
@@ -251,11 +251,9 @@ export class LoginViewController implements ILoginViewController {
 			})
 			.then(() => logins.loginComplete()).then(() => {
 			// don't wait for it, just invoke
-			if (isApp()) {
-				import("../native/common/FileApp")
-					.then(({fileApp}) => fileApp.clearFileData())
-					.catch((e) => console.log("Failed to clean file data", e))
-			}
+			import("../native/common/FileApp")
```
Comment (charlag) [inline src/login/LoginViewController.js:254]: nope, don't always load fileApp, we don't want to import it on web

Comment (johnbotris) [reply src/login/LoginViewController.js:254]: oh yeah...

**Comment by johnbotris** [comment]
I've realised that this isn't granular enough - it needs to be handled on a per window basis

File: src/desktop/DesktopUtils.js [hunk 3]
```diff
@@ -162,6 +161,8 @@ function checkForAdminStatus(): Promise<boolean> {
 }
 
 function getLockFilePath() {
+	// don't get temp dir path from DesktopDownloadManager because we want the lockfile to persist while the app is open,
```
Comment (charlag) [inline src/desktop/DesktopUtils.js:164]: This is unclear to me:
 - Is temp dir for temp files the same dir as the one with lock file? if yes, it shouldn't be
 - Why getting it from DownloadManager should delete the directory? doesn't it create it instead?

File: src/desktop/DesktopMain.js [hunk 4]
```diff
@@ -82,6 +82,14 @@ if (process.argv.indexOf("-r") !== -1) {
 	            .then(() => app.exit(0))
 	            .catch(() => app.exit(1))
 } else {
+	startupInstance()
+}
+
+function startupInstance() {
+	// Delete the temp directory on startup, because we may not always be able to do it on shutdown
+	// we want to await cause the next call will create it again
```
Comment (charlag) [inline src/desktop/DesktopMain.js:90]: "we want to await" but no await is confusing

Comment (johnbotris) [reply src/desktop/DesktopMain.js:90]: oops old comment made obsolete with changes, will delete

File: src/desktop/DesktopUtils.js (hunk 3 above)
Comment (johnbotris) [reply src/desktop/DesktopUtils.js:164]: 1: no it's not, I accidentally made it be when I first did the changes, and realised what I'd done so I wrote a comment to explain the issue
2: getting the directory from the download manager doesn't destroy the directory, but the directory path that the download manager returns will be destroyed at some point during the app lifecycle, so the lockfile goes in `/tmp` and the download manager always gives `/tmp/tutanota/<whatever trailing path>` (basically as it already was)

**Comment by charlag** [comment]
Why does it have to be per-window? because the other window might delete files we need at the moment?
Yeah, on mobile we don't really have thiss...

**Comment by johnbotris** [comment]
> Why does it have to be per-window? because the other window might delete files we need at the moment?

yeah basically, clearFileData get's called during post-login so if we want that to also happen on desktop we'd need to take care of that

**Comment by charlag** [comment]
I don't know what's best, create temp directories per user?..
On the other hand, does it hurt if we delete a file there? it's just a cache, right?

**Comment by johnbotris** [comment]
> it's just a cache, right?

this is a good point, maybe it doesn't matter

**Comment by charlag** [comment]
So if you clean up the comments I would just merge it

**Comment by charlag** [comment]
Thanks!

**Pull request merged**